[Type checker] Guard against recursing into non-single-expression clo…

rudkx opened the pull request:
…sures.

In diagnoseUnintendedOptionalBehavior, guard against recursing into
non-single-expression closures. If we recurse into these, we can end
up hitting malformed ASTs that we do not expect. These happen when we
successfully type-check the code outside of the closure, but the
closure body has errors.

There are potentially other similar issues in other miscellaneous
diagnostics walks.

I opened https://bugs.swift.org/browse/SR-5758 to remind us to review
these and ensure they do not have similar issues.

rudkx commented:
@swift-ci Please smoke test

xedin reviewed (approved):
LGTM!

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@benlangmuir I suspect there may be a better fix here where the code completion expressions aren't treated as errors in the type checker so the change to the code completion code isn't necessary. That would be predicated on the code completion code being able to handle arbitrary invalid decks and null types though. Is it in theory?

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@swift-ci Please test source compatibility

slavapestov commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1269,8 +1270,13 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
       break;
 
     TC.typeCheckDecl(SubDecl, /*isFirstPass*/false);
+    if (SubDecl->isInvalid())
```
I'm not a fan of introducing new usages of the 'invalid' bit.

Also consider a case where SubDecl is not invalid, but one of it's child decls is (and maybe that child decl has a child with a null type somewhere...)

I think declaration checking should leave things in a state where everything has a type, at least. What do you think?

rudkx replied:
I agree that it is more principled to never leave null types around, but I think it still makes sense to communicate the error out and not do e.g. the misc. diagnostics that normally do not do when type checking fails, and the only way that I currently know of to do that is check `isInvalid()` on the decl, or try to dig around for an ErrorType in the expression. For example for the test case from `SR-5666` (which I'm going to add to the PR) we end up with this. We end up crashing on the string_literal_exprs here:

```
(pattern_binding_decl
  (pattern_named type='<<error type>>' 'b')
  (interpolated_string_literal_expr type='<<error type>>' location=/tmp/ppu.swift:8:15 range=[/tmp/ppu.swift:8:15 - line:8:15]
    (string_literal_expr type='<null>' encoding=utf8 value="" builtin_initializer=**NULL** initializer=**NULL**)
    (paren_expr type='<null>'
      (declref_expr type='<<error type>>' location=/tmp/ppu.swift:8:18 range=[/tmp/ppu.swift:8:18 - line:8:18] decl=ppu.(file).testSR5666(cs:).explicit closure discriminator=0.a@/tmp/ppu.swift:7:11 function_ref=unapplied))
    (string_literal_expr type='<null>' encoding=utf8 value="" builtin_initializer=**NULL** initializer=**NULL**)))
```

benlangmuir commented:
> I suspect there may be a better fix here where the code completion expressions aren't treated as errors in the type checker so the change to the code completion code isn't necessary. 

@rudkx I'm not sure what you mean, do you mean `CodeCompletionExpr` specifically, or all things that might include code-completion?

> That would be predicated on the code completion code being able to handle arbitrary invalid decks and null types though. Is it in theory?

Shouldn't we get `ErrorType` not null types if things go wrong during type-checking?  Invalid decls should be fine, and we are willing to add null-checks as needed (and we've done this in the past, so there are checks littered around), but I don't like that the current model exposes multiple failure modes to us.  This is one of the things I already dislike about code-completion driving the type-checking of the "context": if anything gets missed we can end up with a mix of `nullptr` and `ErrorType`.

rudkx commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1269,8 +1270,13 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
-  
+
+  if (hadError)
+    return nullptr;
```
It looks like when processing returns we create a new AST node, but it's the only one I see in a quick review:

```
return new (TC.Context) FailStmt(RS->getReturnLoc(), nilExpr->getLoc(), RS->isImplicit());
```

rudkx commented:
> @rudkx I'm not sure what you mean, do you mean CodeCompletionExpr specifically, or all things that might include code-completion?

I meant `CodeCompletionExpr` specifically, but I don't know enough about how code completion works to know whether I should be concerned with other things.

> Shouldn't we get ErrorType not null types if things go wrong during type-checking? Invalid decls should be fine, and we are willing to add null-checks as needed (and we've done this in the past, so there are checks littered around), but I don't like that the current model exposes multiple failure modes to us. This is one of the things I already dislike about code-completion driving the type-checking of the "context": if anything gets missed we can end up with a mix of nullptr and ErrorType.

We don't fill in all types in the expression with `ErrorType`, and I am not sure how disciplined we are in returning `ErrorType` vs. null types on failure.

Commit:
[Type checker] Guard against recursing into non-single-expression closures.

In diagnoseUnintendedOptionalBehavior, guard against recursing into
non-single-expression closures. If we recurse into these, we can end
up hitting malformed ASTs that we do not expect. These happen when we
successfully type-check the code outside of the closure, but the
closure body has errors.

There are potentially other similar issues in other miscellaneous
diagnostics walks.

I opened https://bugs.swift.org/browse/SR-5758 to remind us to review
these and ensure they do not have similar issues.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
Okay, new/improved fix that deals with the core issue here. I haven't reviewed the other misc diagnostics to see if they have the same issue but did open https://bugs.swift.org/browse/SR-5758 to ensure we do that.
